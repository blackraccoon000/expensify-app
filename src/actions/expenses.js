import uuid from "uuid";
import database from "../firebase/firebase"

const addExpense = (expense) => ({
  type:"ADD_EXPENSE",
  expense
})

const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    // console.log(getState())
    const uid = getState().auth.uid
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }

    return database.ref(`users/${uid}/expenses`).push(expense)
    // return database.ref("expenses").push(expense)
      .then((ref)=>{
        dispatch(addExpense({
          id:ref.key,
          ...expense
        }))
      })
  }
}

const removeExpense = ({id=""}={}) => {
  return {
    type:"REMOVE_EXPENSE",
    id
  }
}

const startRemoveExpense = ({id=""}={}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
      dispatch(removeExpense({id}))
    })
  }
}

const editExpense = (id, updates) => ({
  type:"EDIT_EXPENSE",
  id,
  updates
})

const startEditExpense = (id, updates) => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
    .then(()=>{
      dispatch(editExpense(id, updates))
    })
  }
}

const setExpenses = expenses => ({
  type:"SET_EXPENSES",
  expenses
})

const startSetExpenses = () => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once("value")
    .then(snapshot => {
      const expenses = []
      snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
      dispatch(setExpenses(expenses))
    })
  }
}



export { addExpense, removeExpense, editExpense, setExpenses,
  startAddExpense, startRemoveExpense, startEditExpense, startSetExpenses }