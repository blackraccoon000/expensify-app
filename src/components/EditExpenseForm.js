import React from "react"
import moment from "moment"

export default class EditExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      amount: props.amount,
      calenderFocused: false,
      createdAt: moment(),
      description: props.description ,
      error:undefined,
      note: props.note
    }
  }
  onAmountChange = (e) => {
    const amount = e.target.value
    if(!amount || amount.match(/^\d{1,7}(\.\d{0,2})?$/)) {
      this.setState(()=>({ amount }))
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(()=>({ description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(()=>({ note }))
  }

  onSubmit = (e) => {
    e.preventDefault()
    if(!this.state.description || !this.state.amount) {
      this.setState(()=>{ error:"Please provide description and amount."})
    } else {
      this.setState(()=>{ error:undefined })
      this.props.onSubmit({
        description:this.state.description,
        amount: parseFloat(this.state.amount,10),
        createdAt: this.state.createdAt.valueOf(),
        note:this.state.note
      })
    }
  }

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <label>
            Des:
            <input type="text" value={this.state.description} onChange={this.onDescriptionChange}/>
          </label>
          <label>
            Amount:
            <input type="text" value={this.state.amount} onChange={this.onAmountChange}/>
          </label>
          <label>
            Note:
            <textarea value={this.state.note} onChange={this.onNoteChange}></textarea>
          </label>
          <button>Edit Expense</button>
        </form>
      </div>
    )
  }
}