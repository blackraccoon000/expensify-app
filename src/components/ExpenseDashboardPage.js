import React from "react"
import ExpenseList from "./ExpenseList"
import ExpenseListFilters from "./ExpenseListFilters"

const ExpenseDashboardPage = () => {
  return (
  <div>
    This is from my dashboard component
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)}

export default ExpenseDashboardPage