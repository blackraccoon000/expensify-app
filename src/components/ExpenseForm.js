import React from "react"
import moment from "moment"

// import { connect } from "react-redux"
// import { addExpense } from "../actions/expenses"
import { SingleDatePicker } from "react-dates"

export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props)
    this.state = {
      amount:props.expense ? props.expense.amount.toString() : "",
      calenderFocused: false,
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      description: props.expense ? props.expense.description : "",
      error:undefined,
      note: props.expense ? props.expense.note : "",
    }
  }

  onAmountChange = (e) => {
    const amount = e.target.value

    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      // console.log(amount)
      this.setState(()=>({ amount }))
    }
  }
  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(()=>({ createdAt }))
    }
  }
  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(()=>({ description }))
  }
  onFocusChange = ({ focused }) => {
    // console.log(focused)
    this.setState(()=>({ calenderFocused: focused }))
  }
  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(()=>({ note }))
  }
  onSubmit = (e) => {
    e.preventDefault()
    if(!this.state.description || !this.state.amount) {
      this.setState(()=>({ error:"Please provide description and amount." }))
    } else {
      this.setState(()=>({ error:undefined }))
      this.props.onSubmit({
        description:this.state.description,
        amount: parseFloat(this.state.amount,10),
        createdAt: this.state.createdAt.valueOf(),
        note:this.state.note
      })
    }
  }
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <form onSubmit={this.onSubmit}>
          <input
            type="text" autoFocus placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text" placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}/>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={_=>false}
          />
          <textarea
            placeholder="Add a note for your expense (option)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
        {this.state.error && <p>{this.state.error}</p>}
      </div>
    )
  }
}

// const ExpenseForm = ({dispatch}) => (
//   <div>
//     <form
//       onSubmit={e=>{
//         e.preventDefault()
//         dispatch(addExpense({
//           description:e.target.elements[0].value,
//           amount:e.target.elements[1].value
//           }))
//         console.log("created:",e.target.elements[0].value)
//         e.target.elements[0].value = ""
//         e.target.elements[1].value = ""
//         }
//       }
//     >
//       <label>Des:
//         <input type="text" placeholder="Description" autoFocus name="Des"/>
//         <input type="text" placeholder="Amount" autoFocus name="Amo"/>
//       </label>
//         <input type="submit" value="submit"/>
//     </form>
//   </div>
// )

// export default connect()(ExpenseForm)