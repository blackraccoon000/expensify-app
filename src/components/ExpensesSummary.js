import React from "react"
import { connect } from "react-redux"
import { Link } from "react-router-dom"
import selectExpenses from "../selectors/expenses"
import selectExpensesTotal from "../selectors/expenses-total.js"
import numeral from "numeral"

class ExpensesSummary extends React.Component {
  constructor(props) {
    super(props)
  }
  render() {
    // Count
    const expenseCount = this.props.expenseCount
    const allExpenseCount = this.props.allExpenseCount
    const hiddenExpensesCount = allExpenseCount - expenseCount

    // Word
    const expenseWord = expenseCount === 1 ? "expense" : "expenses"
    const allExpenseWord = allExpenseCount === 1 ? "expense" : "expenses"
    const hiddenExpensesCountWord = hiddenExpensesCount === 1 ? "expense" : "expenses"

    // Total
    const expensesTotal = this.props.expensesTotal
    const allExpensesTotal = this.props.allExpensesTotal
    const hiddenExpensesTotal = allExpensesTotal - expensesTotal

    // Format
    const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00")
    const formattedAllExpensesTotal = numeral(allExpensesTotal / 100).format("$0,0.00")
    const formattedHiddenExpensesTotal = numeral(hiddenExpensesTotal / 100).format("$0,0.00")

    // Message
    const viewingMessage = <h1 className="page-header__title">Viewing <span>{this.props.expenseCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span></h1>
    const allMessage = <h2 className="page-header__subTitle">All {allExpenseWord} <span>{allExpenseCount}</span> {allExpenseWord} totalling <span>{formattedAllExpensesTotal}</span></h2>
    const hiddenMessage = <h2 className="page-header__subTitle">Hidden {hiddenExpensesCountWord} <span>{hiddenExpensesCount}</span> {hiddenExpensesCountWord} totalling <span>{formattedHiddenExpensesTotal}</span> (All {formattedAllExpensesTotal})</h2>

    return (
      <div className="page-header">
        <div className="content-container">
          {expenseCount ? viewingMessage : <h2 className="page-header__alert">No Viewing expenses.</h2>}
          {hiddenExpensesCount ?  hiddenMessage : undefined}

          <div className="page-header__actions">
            <Link className="button" to="/create">Add Expense</Link>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)
  const allExpenses = selectExpenses(state.expenses)
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
    allExpenseCount: allExpenses.length,
    allExpensesTotal: selectExpensesTotal(allExpenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
export { ExpensesSummary }