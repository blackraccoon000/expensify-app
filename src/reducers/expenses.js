const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState,action) => {
  const { type, expense, id:actionId, updates } = action
  switch (type) {
    case "ADD_EXPENSE":
      return [
        ...state,
        expense
      ]
    case "REMOVE_EXPENSE":
      return state.filter(({id}) => id !== actionId)
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if(expense.id === actionId) {
          return {
            ...expense,
            ...updates
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}

export default expensesReducer