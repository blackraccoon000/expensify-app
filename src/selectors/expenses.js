import moment from "moment"

// --> Ex: getVisibleExpenses(expenses,filters)
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  const filterExpenses = expenses.filter(expense => {
    // const startDateMatch = typeof startDate !== "number" || expense.createdAt >= startDate
    // const endDateMatch = typeof endDate !== "number" || expense.createdAt <= endDate
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment , "day") : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment , "day") : true
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
    return startDateMatch && endDateMatch && textMatch
  })
  const sortExpenses = filterExpenses.sort((a,b) => {
    if(sortBy === "date") {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === "amount") {
      return a.amount < b.amount ? 1 : -1
    }
  })
  return sortExpenses
}

export default getVisibleExpenses