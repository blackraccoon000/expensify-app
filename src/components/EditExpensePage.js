import React from "react"
import { connect } from "react-redux"
import { editExpense,startRemoveExpense } from "../actions/expenses"
import ExpenseForm from "./ExpenseForm"

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props)
  }
  onRemove= _ =>{
    this.props.startRemoveExpense({id:this.props.expense.id})
    this.props.history.push("/")
  }
  onSubmit= updates =>{
    this.props.editExpense(this.props.expense.id,updates)
    this.props.history.push("/")
  }
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onRemove}
        >remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state,props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch,props) => ({
  editExpense: (id,updates) => dispatch(editExpense(id,updates)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)