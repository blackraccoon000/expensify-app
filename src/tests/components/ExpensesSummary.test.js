import React from "react"
import { shallow } from "enzyme"
import { ExpensesSummary } from "../../components/ExpensesSummary"
import selectExpensesTotal from "../../selectors/expenses-total"
import expenses from "../fixtures/expenses"

let wrapper

beforeEach(()=>{
  wrapper = shallow(<ExpensesSummary/>)
})

it("should render ExpensesSummary",()=>{
  wrapper = shallow(<ExpensesSummary/>)
  expect(wrapper).toMatchSnapshot()
})

it("should render count 10 total 10000 ExpensesSummary",()=>{
  wrapper.setProps({
    expenseCount:10,
    expensesTotal:10000
  })
  // wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={100}/>)
  expect(wrapper).toMatchSnapshot()
})

it("should render count 10 total 10000 ExpensesSummary",()=>{
  // wrapper = shallow(<ExpensesSummary expenseCount={10} expensesTotal={100000}/>)

  wrapper.setProps({
    expenseCount:expenses.length,
    expensesTotal:selectExpensesTotal(expenses)
  })
  expect(wrapper).toMatchSnapshot()
})

