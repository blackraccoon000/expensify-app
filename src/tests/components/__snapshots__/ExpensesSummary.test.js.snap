// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling 
    $0.00
  </h1>
</div>
`;

exports[`should render count 10 total 10000 ExpensesSummary 1`] = `
<div>
  <h1>
    Viewing 
    10
     
    expenses
     totalling 
    $100.00
  </h1>
</div>
`;

exports[`should render count 10 total 10000 ExpensesSummary 2`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $1,141.95
  </h1>
</div>
`;
